File: neet_code_150_cpp/greedy/partition_label.cpp
#include <iostream>
#include <vector>
#include <string>

class Solution {
    public:
        std::vector<int> partitionLabels(std::string s) {
            return std::vector{1};
        }
};

class SolutionAns {
public:
    std::vector<int> partitionLabels(std::string s) {
        int n = s.size();
        // その文字が最後に何番目にあったを記録
        std::vector<int> lastIndex(26);
        for (int i = 0; i < n; i++) {
            lastIndex[s[i] - 'a'] = i;
        }
        
        int size = 0;
        int end = 0;
        std::vector<int> result;
        // ここまで理解した
        
        for (int i = 0; i < n; i++) {
            size++;
            // constantly checking for further indices if possible
            end = std::max(end, lastIndex[s[i] - 'a']);
            if (i == end) {
                result.push_back(size);
                size = 0;
            }
        }
        
        return result;
    }
};

int main(void) {
    auto case_1 = std::string{"ababcbacadefegdehijhklij"};
    // [9,7,8]
    auto case_2 = std::string{"eccbbbbdec"};
    // [10]

    SolutionAns s_ans;

    // case_1
    auto res_1 = s_ans.partitionLabels(case_1);
    for(auto &x: res_1) {
        std::cout << x << " ";
    }
    std::cout << std::endl;
    // case_2
    auto res_2 = s_ans.partitionLabels(case_2);
    for(auto &x: res_2) {
        std::cout << x << " ";
    }
    std::cout << std::endl;
}

File: neet_code_150/src/greedy/partition_label.rs
// C++の模範解答より
struct Solution {}
impl Solution {
    pub fn partition_labels(s: String) -> Vec<i32> {
        // その文字が最後に何番目にあったかを記録
        let mut last_idx = vec![0; 26];
        for (idx, c) in s.chars().enumerate() {
            last_idx[c as usize - 'a' as usize] = idx;
        }

        let (mut size , mut end) = (0, 0);
        let mut result = vec![];

        for (idx, c) in s.chars().enumerate() {
            size += 1;
            end = std::cmp::max(end, last_idx[c as usize - 'a' as usize]);
            if idx == end {
                result.push(size);
                size = 0;
            }
        }

        result
    }
}

fn main() {
    let case_1 = "ababcbacadefegdehijhklij".to_string();
    // [9, 7, 8]
    let case_2 = "eccbbbbdec".to_string();
    // [10]

    println!("case_1: {:?}", Solution::partition_labels(case_1));
    println!("case_1: {:?}", Solution::partition_labels(case_2));
}
